File: api/routers/monte_carlo.py
"""
api/routers/monte_carlo.py — R4: Server-side Monte Carlo Tolerance Analysis

POST /api/monte-carlo
  Runs N samples of the full DAG physics model with uniform parameter variation.
  Returns per-KPI distributions (mean, std, P10/P50/P90, pass rate, histogram values).

Why server-side:
  - Full DAG model accuracy (vs simplified TS computeAll)
  - N=500 takes ~1-2s in Python (3-5ms/eval × 500 = 1.5-2.5s)
  - Client gets single response; no streaming overhead
"""
import time, math, random
from fastapi import APIRouter
from pydantic import BaseModel
from typing import List, Dict, Optional, Any

from .dag_analysis import DAGRequest
from ._physics import PARAM_META, KPI_META, eval_kpis, patch_request, KPI_TO_TARGET_KEY

router = APIRouter()


# ── Request ───────────────────────────────────────────────────────────────────

class MCRequest(BaseModel):
    bike:         DAGRequest
    tolerances:   Dict[str, float]           # param_id → absolute ±tolerance
    n_samples:    int = 500
    seed:         Optional[int] = None
    targets:      Optional[Dict[str, Any]] = None  # {targetKey: {enabled, lo, hi}}


# ── Response ──────────────────────────────────────────────────────────────────

class KPIStatsOut(BaseModel):
    id:        str
    label:     str
    unit:      str
    values:    List[float]   # all N samples (for histogram rendering)
    mean:      float
    std:       float
    p10:       float
    p50:       float
    p90:       float
    pass_rate: Optional[float]  # None if no target set
    nominal:   float
    target_lo: Optional[float]
    target_hi: Optional[float]

class MCResponse(BaseModel):
    n:                  int
    elapsed_ms:         float
    overall_pass_rate:  float
    kpis:               List[KPIStatsOut]


# ── Helpers ───────────────────────────────────────────────────────────────────

def _percentile(sorted_vals: list, p: float) -> float:
    if not sorted_vals:
        return 0.0
    idx = (p / 100.0) * (len(sorted_vals) - 1)
    lo  = int(idx)
    hi  = min(int(math.ceil(idx)), len(sorted_vals) - 1)
    return sorted_vals[lo] + (sorted_vals[hi] - sorted_vals[lo]) * (idx - lo)


# ── Endpoint ──────────────────────────────────────────────────────────────────

@router.post("/monte-carlo", response_model=MCResponse)
def monte_carlo(req: MCRequest) -> MCResponse:
    t0 = time.perf_counter()

    if req.seed is not None:
        random.seed(req.seed)

    n       = max(1, min(req.n_samples, 5000))   # cap at 5000
    targets = req.targets or {}

    # Enabled params with non-zero tolerance
    enabled = [
        p for p in PARAM_META
        if req.tolerances.get(p['id'], 0.0) > 0
    ]

    # Baseline nominal values
    baseline_kpis = eval_kpis(req.bike)

    # Accumulate per-KPI sample arrays
    kpi_arrays: dict[str, list[float]] = {k['id']: [] for k in KPI_META}

    for _ in range(n):
        # Perturb each enabled param uniformly within ±tolerance
        perturbed = req.bike
        for p in enabled:
            tol = req.tolerances[p['id']]
            v   = getattr(req.bike, p['field'], None)
            if v is None:
                continue
            sampled = float(v) + (random.random() * 2.0 - 1.0) * tol
            perturbed = patch_request(perturbed, p['id'], sampled)

        kpis = eval_kpis(perturbed)
        for k in KPI_META:
            kpi_arrays[k['id']].append(kpis[k['id']])

    # Compute stats
    kpi_stats: list[KPIStatsOut] = []
    for km in KPI_META:
        kpi_id = km['id']
        vals   = kpi_arrays[kpi_id]
        if not vals:
            continue
        sorted_v = sorted(vals)
        mean     = sum(vals) / len(vals)
        variance = sum((v - mean) ** 2 for v in vals) / len(vals)
        std      = math.sqrt(variance)

        # Target range
        tkey      = KPI_TO_TARGET_KEY.get(kpi_id)
        t         = targets.get(tkey) if tkey else None
        target_lo = float(t['lo']) if t and t.get('enabled') else None
        target_hi = float(t['hi']) if t and t.get('enabled') else None
        pass_rate: Optional[float] = None
        if target_lo is not None and target_hi is not None:
            pass_rate = sum(1 for v in vals if target_lo <= v <= target_hi) / len(vals)

        kpi_stats.append(KPIStatsOut(
            id        = kpi_id,
            label     = km['label'],
            unit      = km['unit'],
            values    = [round(v, 4) for v in vals],
            mean      = round(mean, 4),
            std       = round(std, 4),
            p10       = round(_percentile(sorted_v, 10), 4),
            p50       = round(_percentile(sorted_v, 50), 4),
            p90       = round(_percentile(sorted_v, 90), 4),
            pass_rate = round(pass_rate, 4) if pass_rate is not None else None,
            nominal   = round(baseline_kpis[kpi_id], 4),
            target_lo = target_lo,
            target_hi = target_hi,
        ))

    # Overall: fraction of samples where ALL targeted KPIs pass simultaneously
    targeted_ids = [km['id'] for km in KPI_META if km['id'] in kpi_arrays
                    and KPI_TO_TARGET_KEY.get(km['id']) and
                    targets.get(KPI_TO_TARGET_KEY[km['id']], {}).get('enabled')]
    overall_pass_rate = 1.0
    if targeted_ids and n > 0:
        pass_count = 0
        for i in range(n):
            all_pass = True
            for kid in targeted_ids:
                tkey = KPI_TO_TARGET_KEY[kid]
                t    = targets.get(tkey, {})
                v    = kpi_arrays[kid][i]
                if v < t['lo'] or v > t['hi']:
                    all_pass = False
                    break
            if all_pass:
                pass_count += 1
        overall_pass_rate = pass_count / n

    elapsed_ms = (time.perf_counter() - t0) * 1000.0

    return MCResponse(
        n                 = n,
        elapsed_ms        = round(elapsed_ms, 1),
        overall_pass_rate = round(overall_pass_rate, 4),
        kpis              = kpi_stats,
    )

File: chassis-workbench/src/engine/monteCarlo.ts
/**
 * monteCarlo.ts — R4: Monte Carlo Tolerance Analysis
 *
 * Samples each input parameter uniformly within ±tolerance (manufacturing spread),
 * runs computeAll() N times, and computes per-KPI distributions.
 *
 * Tells engineers: "Given real-world build variation, what fraction of bikes
 * will hit their performance targets?" — Cpk thinking for chassis dynamics.
 */

import { ComputeAllInput, TargetConfig } from './types';
import { computeAll } from './computeAll';
import { SENS_PARAMS, SENS_KPIS } from './sensitivity';

// ── MCParam ──────────────────────────────────────────────────────────────────

export interface MCParam {
  id:         string;
  label:      string;
  unit:       string;
  group:      string;
  nominal:    number;
  tolerance:  number;   // absolute ±  (manufacturing spread, 1-sigma or max)
  enabled:    boolean;
  getValue:   (i: ComputeAllInput) => number;
  patchValue: (i: ComputeAllInput, v: number) => ComputeAllInput;
}

// Physics-informed default tolerances (1-sigma manufacturing spread)
const DEFAULT_TOL: Record<string, number> = {
  headAngle:    0.5,    // ±0.5° — frame-welding jig tolerance
  forkOffset:   2,      // ±2 mm — triple-clamp machining
  forkLength:   3,      // ±3 mm — fork tube cut length
  wheelbase:    5,      // ±5 mm — frame assembly stack-up
  swingarmLen:  3,      // ±3 mm — swingarm pivot to axle
  pivotHeight:  2,      // ±2 mm — pivot boss machining
  pivotX:       3,      // ±3 mm — pivot boss X location
  springRateF:  2,      // ±2 N/mm — spring winding tolerance
  springRateR:  2,      // ±2 N/mm
  motionRatioF: 0.02,   // ±0.02 — linkage arm length variation
  motionRatioR: 0.02,
  unsprungF:    0.3,    // ±0.3 kg — wheel/brake component variation
  unsprungR:    0.3,
  frontSprocket:1,      // ±1 tooth (discrete swap)
  rearSprocket: 2,      // ±2 teeth (discrete swap)
};

export function buildMCParams(input: ComputeAllInput): MCParam[] {
  return SENS_PARAMS.map(p => ({
    id:         p.id,
    label:      p.label,
    unit:       p.unit,
    group:      p.group,
    nominal:    p.getValue(input),
    tolerance:  DEFAULT_TOL[p.id] ?? Math.max(Math.abs(p.getValue(input)) * 0.02, 0.01),
    enabled:    true,
    getValue:   p.getValue,
    patchValue: p.patchValue,
  }));
}

// ── KPI target key map ───────────────────────────────────────────────────────

type TCKey = keyof TargetConfig;

const KPITC_KEY: Record<string, TCKey> = {
  trail:     'trail',
  frontPct:  'frontPercent',
  cogH:      'cogHeight',
  antiSquat: 'antiSquatPercent',
  natFreqF:  'natFreqFront',
  natFreqR:  'natFreqRear',
  sagF:      'sagPercentFront',
  sagR:      'sagPercentRear',
  wrF:       'wheelRateFront',
  wrR:       'wheelRateRear',
};

// ── Output types ─────────────────────────────────────────────────────────────

export interface KPIStats {
  id:        string;
  label:     string;
  unit:      string;
  values:    number[];   // all N samples (kept for histogram rendering)
  sorted:    number[];   // pre-sorted
  mean:      number;
  std:       number;
  p10:       number;
  p50:       number;
  p90:       number;
  passRate:  number;     // 0–1 (NaN if no target set)
  nominal:   number;     // baseline design value
  targetLo:  number | null;
  targetHi:  number | null;
}

export interface MCResult {
  n:               number;
  kpis:            KPIStats[];
  elapsed:         number;
  overallPassRate: number;   // fraction of samples ALL targeted KPIs pass simultaneously
}

// ── Helpers ──────────────────────────────────────────────────────────────────

function pct(sorted: number[], p: number): number {
  if (sorted.length === 0) return 0;
  const idx = (p / 100) * (sorted.length - 1);
  const lo  = Math.floor(idx);
  const hi  = Math.min(Math.ceil(idx), sorted.length - 1);
  return sorted[lo] + (sorted[hi] - sorted[lo]) * (idx - lo);
}

// ── Main runner ──────────────────────────────────────────────────────────────

/**
 * Run Monte Carlo simulation.
 * Yields progress every BATCH samples via onProgress (called from async loop).
 * stopRef.current = true halts early.
 */
export async function runMonteCarlo(
  base:       ComputeAllInput,
  params:     MCParam[],
  targets:    TargetConfig,
  n:          number,
  onProgress: (done: number, total: number) => void,
  stopRef:    { current: boolean },
): Promise<MCResult> {
  const t0 = performance.now();
  const BATCH = 25;
  const enabled = params.filter(p => p.enabled && p.tolerance > 0);

  // Accumulate raw values per KPI
  const kpiArrays: number[][] = SENS_KPIS.map(() => []);

  for (let i = 0; i < n; i++) {
    if (stopRef.current) break;

    // Uniform sample: nominal ± tol
    let inp = base;
    for (const p of enabled) {
      const v = p.nominal + (Math.random() * 2 - 1) * p.tolerance;
      inp = p.patchValue(inp, v);
    }

    const r = computeAll(inp);
    SENS_KPIS.forEach((kpi, j) => kpiArrays[j].push(kpi.getValue(r, inp)));

    if ((i + 1) % BATCH === 0) {
      onProgress(i + 1, n);
      await new Promise(res => setTimeout(res, 0));
    }
  }

  const actualN  = kpiArrays[0].length;
  const baseRes  = computeAll(base);

  const kpis: KPIStats[] = SENS_KPIS.map((kpi, j) => {
    const vals   = kpiArrays[j];
    const sorted = [...vals].sort((a, b) => a - b);
    const mean   = vals.reduce((s, v) => s + v, 0) / (vals.length || 1);
    const std    = Math.sqrt(vals.reduce((s, v) => s + (v - mean) ** 2, 0) / (vals.length || 1));
    const tKey   = KPITC_KEY[kpi.id];
    const tr     = tKey ? targets[tKey] : null;
    const targetLo = (tr?.enabled) ? tr.lo : null;
    const targetHi = (tr?.enabled) ? tr.hi : null;
    const passRate = (targetLo !== null && targetHi !== null)
      ? vals.filter(v => v >= targetLo! && v <= targetHi!).length / (vals.length || 1)
      : NaN;
    const nominal = kpi.getValue(baseRes, base);

    return {
      id: kpi.id, label: kpi.label, unit: kpi.unit, values: vals, sorted,
      mean, std,
      p10: pct(sorted, 10),
      p50: pct(sorted, 50),
      p90: pct(sorted, 90),
      passRate, nominal, targetLo, targetHi,
    };
  });

  // Overall: fraction of samples where EVERY targeted KPI passes simultaneously
  const targeted = kpis.filter(k => k.targetLo !== null);
  let overallPassRate = 1;
  if (targeted.length > 0 && actualN > 0) {
    let passCount = 0;
    for (let i = 0; i < actualN; i++) {
      if (targeted.every(k => k.values[i] >= k.targetLo! && k.values[i] <= k.targetHi!)) {
        passCount++;
      }
    }
    overallPassRate = passCount / actualN;
  }

  return { n: actualN, kpis, elapsed: performance.now() - t0, overallPassRate };
}
